# Ejercicio 7

> Por: Mapachana

**Toma tu número n de la lista publicada para el ejercicio 3. Sea d elprimer elemento de la sucesión 5, -7, 9, -11, 13, ... que satisface que el símbolo deJacobi es (d|n) = -1**

In [1]:
n = 36580545945776718558633000960211
n

36580545945776718558633000960211

In [7]:
def jacobi(a, n):
        #assert(n > a > 0 and n%2 == 1)
        t = 1
        while a != 0:
            while a % 2 == 0:
                a /= 2
                r = n % 8
                if r == 3 or r == 5:
                    t = -t
            a, n = n, a
            if a % 4 == n % 4 == 3:
                t = -t
            a %= n
        if n == 1:
            return t
        else:
            return 0

In [13]:
num_base = 5
exp = 2

while True:
    num_suc = pow(-1, exp)*num_base
    print(num_suc)
    
    if jacobi(num_suc, n) == -1:
        print(num_suc)
        break
    
    if exp >= 10:
        break
    
    num_base += 2
    exp += 1
    


5
-7
9
-11
13
13


In [15]:
d = 13
d

13

Luego el primer número de la sucesión que cumple $ \left( \frac{d}{n} \right) =-1$ es 13, y por tanto $d=13$.

In [17]:
import math

**i) Con P = 1, Q = (1-d)/4, define el e.c. $\alpha$ y sus sucesiones de Lucas asociadas.**

In [22]:
P = 1
Q = (1-d)/4

alfa = (P+math.sqrt(pow(P,2)-4*Q))/2
print(alfa)
print(pow(P,2)-4*Q)
print(Q)
print((1+math.sqrt(13))/2)

2.302775637731995
13.0
-3.0
2.302775637731995


Sabemos que $\alpha = \frac{P+\sqrt{P^2-4Q}}{2}$, luego tenemos que $\alpha = \frac{1+\sqrt{13}}{2} =2.302775637731995$

Además, sabemos que $\alpha$ es un entero cuadrático por ser solución de $x^2-x-3=0 \in \mathbb{Z}[x]$.

Entonces, se tiene $\alpha^n = \alpha^{n-1}+3\alpha^{n-2}$.

Por tanto si llamamos $\alpha^{i}=\frac{V_i}{2}+\frac{U_i}{2}\sqrt{13}$ se tiene 
$$U_0=0, \quad V_0=2$$
$$U_1=1, \quad V_1=1$$


La sucesión de Lucas asociada es:

$$V_n=V_{n-1}+3V_{n-2}$$
$$U_n=U_{n-1}+3U_{n-2}$$

**ii) Si n primo ¿Que debería de pasarle a $V_r,U_r$,módulo n? ¿Y a $V_{r/2} ,U_{r/2}$? Calcula los términos $V_{r/2} ,U_{r/2},V_r ,U_r$, módulo n, de las sucesiones de Lucas. ¿ Tu n verifica el TPF para el entero cuadrático $\alpha$ ?**

**iii) Factoriza r = n+1 y para cada factor primo p suyo, calcula $U_{r/p}$ . ¿ Cuál es el rango de Lucas w(n) ?. ¿ Qué deduces sobre la primalidad de tu n ?**